debug/org.eclipse.cdt.debug.ui/src/org/eclipse/cdt/debug/internal/ui/actions/PinViewHandler.java

```diff
@@ -301,19 +218,41 @@ private void updatePinContextColor(DebugContextPinProvider provider) {
 
 		if (imageDesc == null)
 			imageDesc = CDTSharedImages.getImageDescriptor(CDTSharedImages.IMG_VIEW_PIN_ACTION);
-		fAction.setImageDescriptor(imageDesc);
+		image = imageDesc.createImage();
+		toolItem.setImage(image);
 	}
 
-	/* (non-Javadoc)
-	 * @see org.eclipse.debug.ui.contexts.IDebugContextListener#debugContextChanged(org.eclipse.debug.ui.contexts.DebugContextEvent)
-	 */
-	@Override
-	public void debugContextChanged(DebugContextEvent event) {
-		if (fAction != null && !fAction.isChecked()) {
-			final boolean pinnable = PinCloneUtils.isPinnable(fPart, event.getContext());
-			if (pinnable != fAction.isEnabled()) {
-				PlatformUI.getWorkbench().getDisplay().syncExec(() -> fAction.setEnabled(pinnable));
+	private boolean useMultiPinImage(Set<IPinElementHandle> handles) {
+		if (handles.size() <= 1)
+			return false;
+
+		int overlayColor = IPinElementColorDescriptor.UNDEFINED;
+		ImageDescriptor imageDesc = null;
+		for (IPinElementHandle handle : handles) {
+			IPinElementColorDescriptor colorDesc = handle.getPinElementColorDescriptor();
+			if (colorDesc != null) {
+				ImageDescriptor descImageDesc = colorDesc.getToolbarIconDescriptor();
+				if (imageDesc != null && !imageDesc.equals(descImageDesc))
+					return true;
+				imageDesc = descImageDesc;
+
+				int descOverlayColor = colorDesc.getOverlayColor();
+				if (overlayColor != IPinElementColorDescriptor.UNDEFINED && descOverlayColor != overlayColor)
+					return true;
+				overlayColor = descOverlayColor;
 			}
 		}
+
+		return false;
+	}
+
+	@Override
+	public void dispose() {
+		pinned = null;
```

Comment:
Make `pinned` final and `clear()` it instead of setting to null as the setting to null has a bunch of possible NPEs it causes if this handler gets disposed.

I feel I am probably overthinking this, but since in theory a new instance of the handler could be instantiated, the static `pinned` should remain usable.

I have applied this change in the push I did just now.